Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -462,100 +462,80 @@ Most of the frameworks have a parameter to configure the session timeout. This p
-
-#### Remediation
-
--- TODO [Describe the best practices that developers should follow to prevent this issue "Testing 2-Factor Authentication".] --
-
-#### References
+Applications that offer access to sensitive data or critical functions, might require users additionally to re-authenticate with a stronger authentication mechanism. For example, after logging in via biometric authentication (e.g. TouchID) into a banking app, a user might need to do a so called "Step-up Authentication" again through OTP in order to execute a bank transfer.
 
-##### OWASP Mobile Top 10 2016
-* M4 - Insecure Authentication - https://www.owasp.org/index.php/Mobile_Top_10_2016-M4-Insecure_Authentication
+A key advantage of step-up authentication is improved usability for the user. A user is asked to authenticate with the additional factor only when necessary.
 
-##### OWASP MASVS
-* 4.8: "A second factor of authentication exists at the remote endpoint and the 2FA requirement is consistently enforced."
-
-##### CWE
-
--- TODO [Add relevant CWE for "Testing 2-Factor Authentication"] --
-- CWE-312 - Cleartext Storage of Sensitive Information
 
-##### Info
+#### Static Analysis
 
-- [1] Meyer's Recipe for Tomato Soup - http://www.finecooking.com/recipes/meyers-classic-tomato-soup.aspx
-- [2] Another Informational Article - http://www.securityfans.com/informational_article.html
+When server-side source code is available, first identify how a second factor or step-up authentication is used and enforced. Afterwards locate all endpoints with sensitive and privileged information and functions: they are the ones that need to be protected. Prior to accessing any item, the application must make sure the user has already passed 2FA or the step-up authentication and that he is allowed to access the endpoint.
 
-##### Tools
+2FA or step-up authentication shouldn't be implemented from scratch, instead they should be build on top of available libraries that offer this functionality. The libraries used on the server side should be identified and the usage of the available APIs/functions should be verified if they are used accordingly to best practices.
 
--- TODO [Add relevant tools for "Testing 2-Factor Authentication"] --
-* Enjarify - https://github.com/google/enjarify
+For example server side libraries like GoogleAuth<sup>[2]</sup> can be used. Such libraries rely on a widely accepted mechanism of implementing an additional factor by using Time-Based One-Time Password Algorithms (TOTP). TOTP is a cryptographic algorithm that computes a OTP from a shared secret key between the client and server and the current time. The created OTPs are only valid for a short amount of time, usually 30 to 60 seconds.
 
+Instead of using libraries in the server side code, also available cloud solutions can be used like for example:
 
+- Google Authenticator<sup>[2]</sup>
+- Microsoft Authenticator<sup>[3]</sup>
+- Authy<sup>[4]</sup>
```

Comment:
I wonder, should the pentester test whether these solutions are implemented correctly? Maybe for high-risk applications?